### PR TITLE
dvc: require `--external` for external outputs

### LINKED ISSUE #1545: add: from external FS or local external outputs
```
$ dvc add /Volumes/Seagate/storage/data/reddit-May2015_z.tsv.zip
```
It copies the file into the cache, removes the original file in the external drive and creates symlink instead (to the dvc cache).

Based on some offline discussions it confuses people. One of the biggest concerns - after removing DVC project (the entire dir with the code as well as DVC-cache) user sees a symlink to a file which does not exist.

1. Why should DVC work with external files? Is there any meaningful scenario? If not - let's throw an error when adding external files.
2. Even if we have a scenario for (1) let's not touch\remove external files?
3. As a result of the command, I have `reddit-May2015_z.tsv.zip.dvc` file but not the actual file in the workspace (like `reddit-May2015_z.tsv.zip`). It is not clear how to instantiate the file in the workspace?

### PR DESCRIPTION
**Pull request by efiop**
Fixes #1545
Related to #3920

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here. If the CLI API is changed, I have updated [tab completion scripts](https://github.com/iterative/dvc/tree/master/scripts/completion).

https://github.com/iterative/dvc.org/pull/1387
* [x] ❌ I will check DeepSource, CodeClimate, and other sanity checks below. (We consider them recommendatory and don't expect everything to be addressed. Please fix things that actually improve code or fix bugs.)

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/stage/__init__.py [hunk 1]
```diff
@@ -32,6 +32,35 @@
 INVALID_STAGENAME_CHARS = set(string.punctuation) - {"_", "-"}
 
 
+class ExternalOutputsError(DvcException):
+    pass
+
+
+def _check_no_externals(outs):
```
Comment (skshetry) [inline dvc/stage/__init__.py:39]: I think, we should have this check outside on `add()` as it's not really a stage that's external but an output.

File: dvc/stage/__init__.py [hunk 2]
```diff
@@ -32,6 +32,35 @@
...
+    from urllib.parse import urlparse
+
+    # NOTE: preventing users from accidentally using external outputs. See
+    # https://github.com/iterative/dvc/issues/1545 for more details.
+
+    def _is_external(out):
+        # NOTE: if user uses `remote://` notation, he clearly knows what he is
+        # doing and we shouldn't error-out.
```
Comment (skshetry) [inline dvc/stage/__init__.py:47]:
```suggestion
        # NOTE: on `remote://` notation, we shouldn't error-out.
```

Comment (efiop) [reply dvc/stage/__init__.py:47]: @skshetry The new comment doesn't provide any hint on `why` though :slightly_smiling_face:

File: dvc/stage/__init__.py (hunk 1 above)
Comment (efiop) [reply dvc/stage/__init__.py:39]: @skshetry It is also used on `run` so it made sense to put it in one place. We also have some other out-relateted checks in create_stage already, like checking the cycles or output overlaps.

Comment (skshetry) [reply dvc/stage/__init__.py:39]: Ahh, missed the `run`.

Comment (skshetry) [reply dvc/stage/__init__.py:39]: Stage is already bloated, so I didn't want this check here. But, as it's also on `run`, it's okay, I guess.

File: dvc/stage/__init__.py (hunk 2 above)
Comment (skshetry) [reply dvc/stage/__init__.py:47]: @efiop, then let's make it gender-neutral. :smile:

File: dvc/stage/__init__.py (hunk 1 above)
Comment (skshetry) [inline dvc/stage/__init__.py:39]: So, let's move this to `stage.utils` then.

File: dvc/command/run.py [hunk 3]
```diff
@@ -225,6 +226,12 @@ def add_parser(subparsers, parent_parser):
         default=False,
         help=argparse.SUPPRESS,
     )
+    run_parser.add_argument(
+        "--external",
+        action="store_true",
+        default=False,
+        help="Allow outputs that are outside of the DVC project.",
```
Comment (skshetry) [inline dvc/command/run.py:233]: Do we use `project`? Maybe:
```suggestion
        help="Allow outputs that are outside of the DVC repo.",
```

Or, simply `Allow external outputs`?

**Pull request merged**